spec/aws-record/record/dirty_tracking_spec.rb

```diff
@@ -0,0 +1,240 @@
+# Copyright 2015-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License is
+# located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions
+# and limitations under the License.
+
+require 'spec_helper'
+require 'securerandom'
+
+describe Aws::Record::DirtyTracking do
+
+  let(:klass) do
+    Class.new do
+      include(Aws::Record)
+
+      set_table_name(:test_table)
+
+      string_attr(:mykey, hash_key: true)
+      string_attr(:body)
+    end
+  end
+
+  let(:instance) { klass.new }
+
+  let(:stub_client) { Aws::DynamoDB::Client.new(stub_responses: true) }
+
+  describe '#[attribute]_dirty?' do 
+
+    it "should return whether the attribute is dirty or clean" do 
+      expect(instance.mykey_dirty?).to be false
+
+      instance.mykey = SecureRandom.uuid
+      expect(instance.mykey_dirty?).to be true
+    end
+
+    it "should not reflect changes to the original value as dirty" do 
+      instance.mykey = nil
+      expect(instance.mykey_dirty?).to be false
+
+      instance.mykey = SecureRandom.uuid
+      expect(instance.mykey_dirty?).to be true
+
+      instance.mykey = nil
+      expect(instance.mykey_dirty?).to be false
+    end
+  end
+
+  describe '#[attribute]_dirty!' do 
+
+    before(:each) do 
+      instance.mykey = "Alex"
+      instance.clean!
+    end
+
+    it "should mark the attribute as dirty" do 
+      instance.mykey << 'i'
+      expect(instance.dirty?).to be false
+
+      instance.mykey_dirty!
+      expect(instance.dirty?).to be true 
+      expect(instance.mykey_was).to eq 'Alexi'
+      expect(instance.mykey).to eq 'Alexi'
+
+      instance.mykey << 's'
+      expect(instance.mykey_was).to eq 'Alexi'
```

Comment:
This is actually testing `attribute_was` rather than dirty marking. Should pull this into a separate test to make regression identification a bit simpler down the road.